### ISSUE TITLE
New proposed method of loot generation that would hopefully make more sense.

### ISSUE THREAD

**Issue by Asmageddon** [CleverRaven/Cataclysm-DDA#19941]
Namely, generate loot in sets, e.g. a 9mm gun and some ammo, and define specific squares as "loot containers", to which said loot is allocated, instead of each square choosing its loot at random with no regard for other stuff.

Would make ensuring that garages/public works have a set of tools more easily, sensible gun+ammo combinations, etc.

**Comment by keyspace** [comment]
* #15443 - Can we have more firearm spawns that already contain a magazine? (closed)
* #15341 - New mapgen and gunstores (merged)

**Comment by keyspace** [comment]
Reviewed #15341. Seems to fit `subj`.

**Comment by Asmageddon** [comment]
Yeah, it does basically solve the same issue, in a manner. But I think it would make more sense in the long term overall. It'd help generating more well-rounded sets of loots, limiting or increasing variety, ensuring that one tool or another is available, etc. etc. It's not a big deal, but a sensible proposal I think.

**Comment by keyspace** [comment]
> Yeah, it does basically solve the same issue, in a manner.

No, it implements _exactly_ what you are proposing to be implemented, and provides a few examples.

Out of the 147 files in `data/json/mapgen`, 73 have the older `place_items`, whereas 60 have the new `place_loot`.

**Comment by keyspace** [comment]
IMO OK to close.

**Comment by Asmageddon** [comment]
Okay. Out of curiosity how does place_loot work? Also sorry for being a bother.

**Comment by keyspace** [comment]
I think I understand the gist of your proposal better now. See after the second break.

-----

`place_items` takes an item group (historically called `"items"` in mapgen JSON), - along with `"chance"` and `"x"`/`"y"` coordinates (referenced to the mapgen object's `"rows"`), - and places items from that group `"repeat"` times.

`place_loot` can take all of the above, but the items can be specified as an item group (via `"group"`) _or a specific item_ (via `"item"`).

The latter can be used for specific-chance item spawns, e.g. [guaranteed `puller` and `press` in the gunsmith's shop](https://github.com/CleverRaven/Cataclysm-DDA/pull/15341/files#diff-a792873748865a540269a5f2fa79edc9R120) in the referenced PR.

(If the spawned items seem "unrelated", then perhaps a different group should be used. Or, the mapgen object specifies several groups over the same tiles. Either way, it's case-by-case.)

`place_loot` can also take `"ammo"` and `"magazine"` for guns. This covers your example of guns+magazines+ammo, which is a honorary special case, since the ammo type has to be matched.

As for the

> define specific squares as "loot containers"

part, it has only recently been merged in #19793. It is more of a QoL thing for map-makers. It was always there, just "not yet" for `place_loot`.

-----

The description above _seems_ to lack an "all or nothing" item group spawn. This is not true.

In #15341, `place_loot` uses `item_group::items_from()`, the results of which depend on the `subtype` of an item group - `distribution` or `collection`. The latter AFAIK is only used in map extras ATM (see `data/json/map_extra_items.json`) - that's why death scenes with lab coats also have lab ID cards.

Since then, the call has been replaced with `Item_group::add_group_entry()` (in #19658), which should achieve the same result.

-----

**TL;DR:** All parts of proposal already implemented by using `place_loot` (or just `loot`) in mapgen JSON objects (see [MAPGEN.md](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/MAPGEN.md#2616-loot)), and referencing item groups of `collection` type (see [ITEM_SPAWN.md](https://github.com/CleverRaven/Cataclysm-DDA/blob/master/doc/ITEM_SPAWN.md#1-collection-or-distribution)).

**Comment by keyspace** [comment]
P.S. That is, it's implemented code-side. Most of JSON still uses `place_items` and `distribution` groups - they work well enough for most cases.

"Porting" it is a high-cost/low-benefit task, both for implementing and reviewing.
